pkg/watch/manager.go

```diff
@@ -33,330 +36,268 @@ type Manager struct {
+	return nil
 }
 
-// updateManager scans for changes to the watch list and restarts the manager if any are detected
-func (wm *Manager) updateManager() (bool, error) {
-	intent, err := wm.managedKinds.Get()
-	if err != nil {
-		return false, errp.Wrap(err, "error while retrieving managedKinds, not restarting watch manager")
-	}
-	if err := wm.metrics.reportGvkIntentCount(int64(len(intent))); err != nil {
-		log.Error(err, "while reporting gvk intent count metric")
-	}
-	added, removed, changed, err := wm.gatherChanges(intent)
-	if err != nil {
-		return false, errp.Wrap(err, "error gathering watch changes, not restarting watch manager")
-	}
-	started := wm.started.Load().(bool)
-	if started && len(added) == 0 && len(removed) == 0 && len(changed) == 0 {
-		return false, nil
-	}
-	var a, r, c []string
-	for k := range added {
-		a = append(a, k.String())
-	}
-	for k := range removed {
-		r = append(r, k.String())
-	}
-	for k := range changed {
-		a = append(c, k.String())
+func (wm *Manager) close() {
+	log.Info("attempting to stop watch manager...")
+	wm.startedMux.Lock()
+	defer wm.startedMux.Unlock()
+	wm.stopper()
+	log.Info("waiting for watch manager to shut down")
+	if wm.stopped != nil {
+		<-wm.stopped
 	}
-	log.Info("Watcher registry found changes and/or needs restarting", "started", started, "add", a, "remove", r, "change", c)
+	log.Info("watch manager finished shutting down")
+}
 
-	readyToAdd, err := wm.filterPendingResources(added)
-	if err != nil {
-		return false, errp.Wrap(err, "could not filter pending resources, not restarting watch manager")
-	}
+func (wm *Manager) GetManagedGVK() []schema.GroupVersionKind {
+	return wm.managedKinds.GetGVK()
+}
 
-	if started && len(readyToAdd) == 0 && len(removed) == 0 && len(changed) == 0 {
-		log.Info("Only changes are pending additions; not restarting watch manager")
-		return false, nil
-	}
+func (wm *Manager) addWatch(gvk schema.GroupVersionKind) error {
+	wm.watchedMux.Lock()
+	defer wm.watchedMux.Unlock()
+	return wm.doAddWatch(gvk)
+}
 
-	newWatchedKinds := make(map[schema.GroupVersionKind]vitals)
-	for gvk, vitals := range wm.watchedKinds {
-		if _, ok := removed[gvk]; !ok {
-			if newVitals, ok := changed[gvk]; ok {
-				newWatchedKinds[gvk] = newVitals
-			} else {
-				newWatchedKinds[gvk] = vitals
-			}
-		}
-	}
+func (wm *Manager) doAddWatch(gvk schema.GroupVersionKind) error {
+	// lock acquired by caller
 
-	filteredNewWatchedKinds, err := wm.filterPendingResources(newWatchedKinds)
-	if err != nil {
-		return false, errp.Wrap(err, "could not filter new watched kinds, not restarting watch manager")
-	}
-	if len(filteredNewWatchedKinds) != len(newWatchedKinds) {
-		var missing []string
-		for k := range newWatchedKinds {
-			if _, ok := filteredNewWatchedKinds[k]; !ok {
-				missing = append(missing, k.String())
-			}
-		}
-		log.Info("previously watched resources have gone pending, removing them from the watch list", "pending", missing)
+	if _, ok := wm.watchedKinds[gvk]; ok {
```

Comment:
Here is where we could run the cache list and broadcast events to the newly-interested controller.

Reply:
Things get tricky when joining an active informer. A few considerations:
1. `AddWatch` needs to be idempotent, we'd need to differentiate if the watcher is actually new. Not much of a problem.
2. Events can be coming in concurrently to the List() for replay. The informer interface doesn't let us pause its internal processing, so here's some wild ideas:

* We might pause receiving/dispatching events from that particular informer on our end until the replay concludes. We may need to do some max-resource-version based dropping of watch events once we resume, as they will have already reached the informer store.
* We might rework things to add a separate event handler per-registrar. `AddEventHandler` properly joins the stream within the informer. When removing the registrar, we wouldn't be able to remove the handler so we'd need to blackhole any events destined for it.
* We might keep track of incoming watch events during the replay. We could drop replay events from List() if their resourceVersion is older than a watch event we observed.

I like the second or the third the best, what do you think? Or did I overcomplicate this?

Note we only have to worry when multiple registrars watch the same resource kind, aka constraints.

Reply:
Agreed on idempotence.

Do concurrent updates during List matter?

My understanding is that the work queue debounces pending/processing requests, and that all we know is the Namespace/Name (and now GVK) of the resource, so there is no worry about conflicting data. The controller grabs the data from the server as it processes the event.

It seems like if the above paragraph is true, then the worst effect of updates + list is the potential for over-reconciling by one iteration per resource.

Reply:
You're absolutely correct, we don't care about the discrete events and throw away most of the object state in the event in EventPacker in favor of a reference. 👍